byzcoin/contract_deferred.go

```diff
@@ -0,0 +1,404 @@
+	c := &contractDeferred{s: s}
+
+	err := protobuf.Decode(in, &c.DeferredData)
+	if err != nil {
+		return nil, errors.New("couldn't unmarshal instance data: " + err.Error())
+	}
+	return c, nil
+}
+
+func (c *contractDeferred) Spawn(rst ReadOnlyStateTrie, inst Instruction, coins []Coin) (sc []StateChange, cout []Coin, err error) {
+	// This method should do the following:
+	//   1. Parse the input buffer
+	//   2. Compute and store the instruction hashes
+	//   3. Save the data
+	//
+	// Spawn should have those input arguments:
+	//   - proposedTransaction ClientTransaction
+	//   - expireBlockIndex uint64
+	//   - numExecution uint64 (default: 1)
+	cout = coins
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	// 1. Reads and parses the input
+	proposedTransaction := ClientTransaction{}
+	err = protobuf.Decode(inst.Spawn.Args.Search("proposedTransaction"), &proposedTransaction)
+	if err != nil {
+		return nil, nil, errors.New("couldn't decode proposedTransaction: " + err.Error())
+	}
+	expireBlockIndex, err := strconv.ParseUint(string(inst.Spawn.Args.Search("expireBlockIndex")), 10, 64)
+	if err != nil {
+		return nil, nil, errors.New("couldn't convert expireBlockIndex: " + err.Error())
+	}
+	numExecutionBuff := inst.Spawn.Args.Search("NumExecution")
+	NumExecution := defaultNumExecution
+	if len(numExecutionBuff) > 0 {
+		NumExecution, err = strconv.ParseUint(string(numExecutionBuff), 10, 64)
+		if err != nil {
+			return nil, nil, errors.New("couldn't parse NumExecution: " + err.Error())
+		}
+	}
+
+	// 2. Computes the hashes of each instruction and store it
+	hash := make([][]byte, len(proposedTransaction.Instructions))
+	for i, proposedInstruction := range proposedTransaction.Instructions {
+		hash[i] = hashDeferred(proposedInstruction, inst.InstanceID.Slice())
+	}
+
+	// 3. Saves the data
+	data := DeferredData{
+		ProposedTransaction: proposedTransaction,
+		ExpireBlockIndex:    expireBlockIndex,
+		InstructionHashes:   hash,
+		NumExecution:        NumExecution,
+	}
+	var dataBuf []byte
+	dataBuf, err = protobuf.Encode(&data)
+	if err != nil {
+		return nil, nil, errors.New("couldn't encode DeferredData: " + err.Error())
+	}
+
+	sc = append(sc, NewStateChange(Create, inst.DeriveID(""),
+		ContractDeferredID, dataBuf, darcID))
+	return
+}
+
+func (c *contractDeferred) Invoke(rst ReadOnlyStateTrie, inst Instruction, coins []Coin) (sc []StateChange, cout []Coin, err error) {
+	// This method should do the following:
+	//   - Handle the "addProof" invocation
+	//   - Handle the "execProposedTx" invocation
+	//
+	// Invoke:addProof should have the following input argument:
+	//   - identity darc.Identity
+	//   - signature []byte
+	//   - index uint32 (index of the instruction wrt the transaction)
+	err2 := c.checkInvoke(rst, inst.Invoke)
+	if err2 != nil {
+		return nil, nil, errors.New("checks of invoke failed: " + err2.Error())
```

Comment:
this function uses named return values, instead of doing what you have on this line,
you can do 
```
err = c.checkInvoke(...
if err != nil {
  return
}
```

Reply:
By doing so I loose the specific error message added "checks of invoke failed: ...". 

I could do :

```go
if err != nil {
  err = errors.New("checks of invoke failed: " + err2.Error())
  return
} 
```

but that's one line longer and makes the return mechanism a bit more hidden. Is it better?